### PR TITLE
executor: fix a concurrent-access problem caused by accessing a single parser object in session concurrently 

### LINKED ISSUE #24259: panic when reading a partition table by IndexLookUpJoin with IndexReader
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
See the test case below:
```
func (s *testSuite5) TestPartitionTableIndexJoinAndIndexReader(c *C) {
	if israce.RaceEnabled {
		c.Skip("exhaustive types test, skip race test")
	}
	tk := testkit.NewTestKit(c, s.store)
	tk.MustExec("use test")
	tk.MustExec("set @@tidb_partition_prune_mode='dynamic'")
	tk.MustExec(`create table t (a int, b int, key(a))
	partition by range (a) (
		partition p1 values less than (100),
		partition p2 values less than (200),
		partition p3 values less than (300),
		partition p4 values less than (400),
		partition p5 values less than (500),
		partition p6 values less than (600))`)
	tk.MustExec("create table tnormal (a int, b int, key(a), key(b))")
	nRows := 512
	values := make([]string, 0, nRows)
	for i := 0; i < nRows; i++ {
		values = append(values, fmt.Sprintf("(%v, %v)", rand.Intn(nRows), rand.Intn(nRows)))
	}
	tk.MustExec(fmt.Sprintf("insert into t values %v", strings.Join(values, ", ")))
	tk.MustExec(fmt.Sprintf("insert into tnormal values %v", strings.Join(values, ", ")))

	randRange := func() (int, int) {
		a, b := rand.Intn(nRows), rand.Intn(nRows)
		if a > b {
			return b, a
		}
		return a, b
	}
	for i := 0; i < nRows; i++ {
		lb, rb := randRange()
		cond := fmt.Sprintf("(t2.b between %v and %v)", lb, rb)
		result := tk.MustQuery("select t1.a from tnormal t1, tnormal t2 where t1.a=t2.b and " + cond).Sort().Rows()
		fmt.Println(">>>>>>> ", "select /*+ TIDB_INLJ(t1, t2) */ t1.a from t t1, t t2 where t1.a=t2.b and "+cond)
		tk.MustQuery("select /*+ TIDB_INLJ(t1, t2) */ t1.a from t t1, t t2 where t1.a=t2.b and " + cond).Sort().Check(result)
	}
}
```

### 2. What did you expect to see? (Required)
The test can pass.

### 3. What did you see instead (Required)
```
...     "interface conversion: interface {} is nil, not *ast.SelectField\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run.func1\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:454\n" +
...     "runtime.gopanic\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/panic.go:679\n" +
...     "runtime.panicdottypeE\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/iface.go:255\n" +
...     "github.com/pingcap/parser.yyParse\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210330190622-f959a136fc19/parser.go:13853\n" +
...     "github.com/pingcap/parser.(*Parser).Parse\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210330190622-f959a136fc19/yy_parser.go:152\n" +
...     "github.com/pingcap/tidb/session.(*session).ParseSQL\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/session/session.go:1091\n" +
...     "github.com/pingcap/tidb/expression.ParseSimpleExprsWithNames\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/expression/simple_rewriter.go:88\n" +
...     "github.com/pingcap/tidb/planner/core.makePartitionByFnCol\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:793\n" +
...     "github.com/pingcap/tidb/planner/core.(*partitionProcessor).pruneRangePartition\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:745\n" +
...     "github.com/pingcap/tidb/planner/core.PartitionPruning\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/partition_prune.go:40\n" +
...     "github.com/pingcap/tidb/executor.partitionPruning\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3989\n" +
...     "github.com/pingcap/tidb/executor.prunePartitionForInnerExecutor\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:2780\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildIndexReaderForIndexJoin\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3479\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoinInternal\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3211\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoin\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3202\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).fetchInnerResults\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:647\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).handleTask\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:491\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:469\n" +
...     "runtime.goexit\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/asm_amd64.s:1357"
```

### 4. What is your TiDB version? (Required)

```
commit 39762c9d1942478fc8e3fb8e4ad2629fb130c151 (HEAD -> master, origin/master, origin/HEAD)
Author: Shirly <wuxuelian@pingcap.com>
Date:   Fri Apr 23 10:41:54 2021 +0800
```

### LINKED ISSUE #24307: panic when reading a partition table by IndexLookUpJoin with IndexLookUpReader
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Run the test below:
```
func (s *testSuite3) TestPartitionTableIndexJoinIndexLookUp(c *C) {
	if israce.RaceEnabled {
		c.Skip("exhaustive types test, skip race test")
	}
	tk := testkit.NewTestKit(c, s.store)
	tk.MustExec("use test")
	tk.MustExec("set @@tidb_partition_prune_mode='dynamic'")
	tk.MustExec(`create table t (a int, b int, key(a)) partition by hash(a) partitions 4`)
	tk.MustExec("create table tnormal (a int, b int, key(a), key(b))")
	nRows := 512
	values := make([]string, 0, nRows)
	for i := 0; i < nRows; i++ {
		values = append(values, fmt.Sprintf("(%v, %v)", rand.Intn(nRows), rand.Intn(nRows)))
	}
	tk.MustExec(fmt.Sprintf("insert into t values %v", strings.Join(values, ", ")))
	tk.MustExec(fmt.Sprintf("insert into tnormal values %v", strings.Join(values, ", ")))

	randRange := func() (int, int) {
		a, b := rand.Intn(nRows), rand.Intn(nRows)
		if a > b {
			return b, a
		}
		return a, b
	}
	for i := 0; i < nRows; i++ {
		lb, rb := randRange()
		cond := fmt.Sprintf("(t2.b between %v and %v)", lb, rb)
		result := tk.MustQuery("select t1.* from tnormal t1, tnormal t2 use index(a) where t1.a=t2.b and " + cond).Sort().Rows()
		tk.MustQuery("select /*+ TIDB_INLJ(t1, t2) */ t1.* from t t1, t t2 use index(a) where t1.a=t2.b and " + cond).Sort().Check(result)
	}
}
```

### 2. What did you expect to see? (Required)
We can pass the test.

### 3. What did you see instead (Required)
```

distsql_test.go:382:
    tk.MustQuery("select /*+ TIDB_INLJ(t1, t2) */ t1.* from t t1, t t2 use index(a) where t1.a=t2.b and " + cond).Sort().Check(result)
/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:346:
    tk.c.Check(errors.ErrorStack(err), check.Equals, "", comment)
... obtained string = "" +
...     "runtime error: index out of range [0] with length 0\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run.func1\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:454\n" +
...     "runtime.gopanic\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/panic.go:679\n" +
...     "runtime.goPanicIndex\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/panic.go:75\n" +
...     "github.com/pingcap/tidb/expression.ParseSimpleExprsWithNames\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/expression/simple_rewriter.go:99\n" +
...     "github.com/pingcap/tidb/planner/core.generateHashPartitionExpr\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:110\n" +
...     "github.com/pingcap/tidb/planner/core.(*partitionProcessor).findUsedPartitions\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:121\n" +
...     "github.com/pingcap/tidb/planner/core.(*partitionProcessor).pruneHashPartition\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/rule_partition_processor.go:207\n" +
...     "github.com/pingcap/tidb/planner/core.PartitionPruning\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/planner/core/partition_prune.go:38\n" +
...     "github.com/pingcap/tidb/executor.partitionPruning\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3980\n" +
...     "github.com/pingcap/tidb/executor.prunePartitionForInnerExecutor\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:2768\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildIndexLookUpReaderForIndexJoin\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3519\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoinInternal\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3205\n" +
...     "github.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoin\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/builder.go:3194\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).fetchInnerResults\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:647\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).handleTask\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:491\n" +
...     "github.com/pingcap/tidb/executor.(*innerWorker).run\n" +
...     "\t/Users/zhangyuanjia/Workspace/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:469\n" +
...     "runtime.goexit\n" +
...     "\t/Users/zhangyuanjia/Workspace/goroot/go1.13/src/runtime/asm_amd64.s:1357"
... expected string = ""
... sql:select /*+ TIDB_INLJ(t1, t2) */ t1.* from t t1, t t2 use index(a) where t1.a=t2.b and (t2.b between 267 and 369), args:[]
```

### 4. What is your TiDB version? (Required)

Master

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #24259, close #24307 

Problem Summary: executor: fix a concurrent-access problem caused by accessing a single parser in session concurrently 

### What is changed and how it works?

The parser object in `session` may be accessed concurrently by `rule_partition_processor` when there is a `DataSource` accessing a partition table in the inner side of an IndexJoin since the `IndexJoin` may create multiple `DataSource`.

After discussing this problem with @tiancaiamao , we decided to introduce a parser pool to solve it.

### Check List 

Tests 

- Unit test

### Release note 

- executor: fix a concurrent-access problem caused by accessing a single parser in session concurrently

### CODE REVIEW

**Comment by qw4990** [comment]
PTAL @tiancaiamao @wshwsh12

File: session/session.go [hunk 1]
```diff
@@ -194,7 +194,7 @@ type session struct {
 
 	store kv.Storage
 
-	parser *parser.Parser
+	parserPool *sync.Pool
```
Comment (tiancaiamao) [inline session/session.go:197]: When there are too many `sync.Pool` objects, it would impose some side effect on Go GC.

File: session/session.go [hunk 2]
```diff
@@ -1089,9 +1089,12 @@ func (s *session) ParseSQL(ctx context.Context, sql, charset, collation string)
 		defer span1.Finish()
 	}
 	defer trace.StartRegion(ctx, "ParseSQL").End()
-	s.parser.SetSQLMode(s.sessionVars.SQLMode)
-	s.parser.SetParserConfig(s.sessionVars.BuildParserConfig())
-	return s.parser.Parse(sql, charset, collation)
+
+	p := s.parserPool.Get().(*parser.Parser)
+	defer s.parserPool.Put(p)
+	p.SetSQLMode(s.sessionVars.SQLMode)
```
Comment (tiancaiamao) [inline session/session.go:1095]: When the pool object is reused, the `SQLMode` will change the behaviour.

**Comment by tiancaiamao** [comment]
/LGTM

There are some flaws in this solution but I don't have better ideas.

File: session/session.go (hunk 2 above)
Comment (qw4990) [reply session/session.go:1095]: Yes, so we have to set its SQL Mode each time before using it.

**Review by wshwsh12 — APPROVED** [review]
LGTM

**Pull request merged**